sdk-cpp/include/do_download_property.h

```diff
@@ -45,56 +49,48 @@ enum class download_property
     integrity_check_info,           // std::string
     integrity_check_mandatory,      // boolean
     total_size_bytes,               // uint64
-    // Fe and beyond
+
+    // For the COM interface, the following properties are available only in Windows 21H2 (Build Number 22000) and beyond
     disallow_on_cellular,           // bool
     http_custom_auth_headers,       // std::string
 };
 
 class download_property_value
 {
 
+/*
+CDownloadImpl is declared as a friend class because it needs to access the platform-specific native value for download_property_value
+The type of the native value is defined in CDownloadPropertyValueInternal, because DO header files are platform agnostic
+This is so any user of the SDK does not have to worry about supplying platform specific compile definitions to use the SDK
+*/
+friend class details::CDownloadImpl; 
+
 public:
-    using native_type = VARIANT;
     using status_callback_t = std::function<void(download&, download_status&)>;
-
     download_property_value() = default;
 
     explicit download_property_value(const std::string& val);
-    explicit download_property_value(UINT val);
-    explicit download_property_value(UINT64 val);
+    explicit download_property_value(uint32_t val);
+    explicit download_property_value(uint64_t val);
     explicit download_property_value(bool val);
     explicit download_property_value(std::vector<unsigned char>& val); // DODownloadProperty_SecurityContext
 
     explicit download_property_value(const status_callback_t& val);
 
-    ~download_property_value();
-
-    download_property_value(const download_property_value& rhs);
-    download_property_value& operator=(download_property_value copy);
-    download_property_value(download_property_value&& rhs) noexcept;
-
-    friend void swap(download_property_value& first, download_property_value& second) noexcept
-    {
-        std::swap(first._var, second._var);
-        std::swap(first._callback, second._callback);
-    }
+    ~download_property_value() = default;
 
     void as(bool& val) const;
-    void as(UINT& val) const;
-    void as(UINT64& val) const;
+    void as(uint32_t& val) const;
+    void as(uint64_t& val) const;
     void as(std::string& val) const;
 
     void as(status_callback_t& val) const;
     void as(std::vector<unsigned char>& val) const;
 
-    const native_type& native_value() const;
-
 private:
-    native_type _var;
-    status_callback_t _callback;
+    std::unique_ptr<details::CDownloadPropertyValueInternal> _val;
```

Comment:
Actually, not so straightforward to convert to unique_ptr, since it's a unique_ptr class must define a move constructor which forwards _val into the copy, and delete the copy constructor.
Since we this class is used as a return value internally - we must provide do the above.
But at the same time we cannot delete an incomplete type of CDownloadPropertyValImpl stored in the unique ptr if we try to delete the copy constructor

Reply:
Anyways, because this may get refactored in the future, going to use std::shared_ptr

Reply:
Sounds good.